File: lib/piece.rb

class Piece < Errol::Entity
  require_relative './piece/record'
  def self.repository
    Catalogue
  end

  entry_accessor  :id,
                  :title,
                  :sub_heading,
                  :level_overview,
                  :description,
                  :notation_preview,
                  :audio_preview,
                  :cover_image,
                  :print_link,
                  :print_title,
                  :weezic_link,
                  :meta_description,
                  :meta_keywords

  boolean_accessor  :beginner,
                    :intermediate,
                    :advanced,
                    :professional,
                    :solo,
                    :solo_with_accompaniment,
                    :duet,
                    :trio,
                    :quartet,
                    :larger_ensembles,
                    :collection,
                    :piano,
                    :recorder,
                    :flute,
                    :oboe,
                    :clarineo,
                    :clarinet,
                    :bassoon,
                    :saxophone,
                    :trumpet,
                    :violin,
                    :viola,
                    :percussion

  def items
    record.item_records.map{ |r| Item.new r }
  end

  def catalogue_number
    "UD#{record.id}" if record.id
  end

  def product_name
    "#{title} - #{sub_heading}"
  end

  def categories
    [:solo,
    :solo_with_accompaniment,
    :duet,
    :trio,
    :quartet,
    :larger_ensembles].select do |category|
      public_send "#{category}?"
    end
  end

  def instruments
    [:piano,
    :recorder,
    :flute,
    :oboe,
    :clarineo,
    :clarinet,
    :bassoon,
    :saxophone,
    :trumpet,
    :violin,
    :viola,
    :percussion].select do |instrument|
      public_send "#{instrument}?"
    end
  end

  def levels
    [:beginner,
    :intermediate,
    :advanced,
    :professional].select do |level|
      public_send "#{level}?"
    end
  end

end

File: project/apps/catalogue/lib/catalogue/piece.ex
defmodule UM.Catalogue.Piece do
  defstruct [
    id: "", # Integer
    title: "", # String
    sub_heading: "", # String
    level_overview: "", # String
    description: "", # String
    solo: false,
    solo_with_accompaniment: false,
    duet: false,
    trio: false,
    quartet: false,
    larger_ensembles: false,
    collection: false,
    beginner: false,
    intermediate: false,
    advanced: false,
    professional: false,
    piano: false,
    recorder: false,
    flute: false,
    oboe: false,
    clarineo: false,
    clarinet: false,
    bassoon: false,
    saxophone: false,
    trumpet: false,
    violin: false,
    viola: false,
    percussion: false,
    notation_preview: "", # Hash
    audio_preview: "", # Hash
    cover_image: "", # Hash
    print_link: "", # String
    print_title: "", # String
    weezic_link: "", # String
    meta_description: "", # String
    meta_keywords: "", # String
  ]

  def catalogue_number(%{id: id}) do
    "UD#{id}"
  end

  def product_name(%{title: title, sub_heading: sub_heading}) do
    "#{title} - #{sub_heading}"
  end

  def categories(piece) do
    Enum.filter([:solo,
    :solo_with_accompaniment,
    :duet,
    :trio,
    :quartet,
    :larger_ensembles], fn (key) ->
      {:ok, bool} = Map.fetch(piece, key)
      bool
    end)
  end

  def instruments(piece) do
    Enum.filter([:piano,
    :recorder,
    :flute,
    :oboe,
    :clarineo,
    :clarinet,
    :bassoon,
    :saxophone,
    :trumpet,
    :violin,
    :viola,
    :percussion], fn (key) ->
      {:ok, bool} = Map.fetch(piece, key)
      bool
    end)
  end

  def levels(piece) do
    Enum.filter([:beginner,
    :intermediate,
    :advanced,
    :professional], fn (key) ->
      {:ok, bool} = Map.fetch(piece, key)
      bool
    end)
  end
end
